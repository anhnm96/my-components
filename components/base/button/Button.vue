<script lang="ts" setup>
const props = withDefaults(
  defineProps<{
    loading?: boolean
    loadingMsg?: string
    success?: boolean
    successMsg?: string
    contentClass?: string
  }>(),
  {
    loading: false,
    loadingMsg: 'processing, wait...',
    success: false,
  },
)
const emit = defineEmits<{
  (e: 'click', ev: MouseEvent): void
}>()

const btnRef = ref<HTMLButtonElement>()
const isBtnUninteractive = computed(() => props.loading || props.success)
function click(event: MouseEvent) {
  const isBtnDisabled = btnRef.value?.getAttribute('aria-disabled') === 'true'
  if (isBtnDisabled || isBtnUninteractive.value) return
  emit('click', event)
}
</script>

<template>
  <button
    ref="btnRef"
    class="btn relative"
    :class="[isBtnUninteractive && '!pointer-events-none']"
    @click="click"
  >
    <span :class="['inline-flex items-center', contentClass, isBtnUninteractive && 'invisible']">
      <slot />
    </span>
    <Transition v-if="isBtnUninteractive" name="fade" mode="out-in">
      <div
        v-if="loading"
        class="flex absolute left-1/2 top-1/2 -translate-x-1/2 -translate-y-1/2"
      >
        <span v-if="loadingMsg" class="sr-only" aria-live="assertive">
          {{ loadingMsg }}
        </span>
        <BaseSpinner />
      </div>
      <div
        v-else-if="success"
        class="flex absolute left-1/2 top-1/2 -translate-x-1/2 -translate-y-1/2"
      >
        <span v-if="successMsg" class="sr-only" aria-live="assertive">
          {{ successMsg }}
        </span>
        <Icon name="ic:baseline-check" />
      </div>
    </Transition>
  </button>
</template>

<style scoped>
button:disabled,
button[aria-disabled='true'] {
  @apply cursor-not-allowed opacity-70;
}
</style>
