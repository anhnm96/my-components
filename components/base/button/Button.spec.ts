import type { VueWrapper } from '@vue/test-utils'
import { shallowMount } from '@vue/test-utils'
import Button from './Button.vue'

describe('Button.vue', () => {
  it('render correctly', () => {
    const wrapper = shallowMount(Button)
    expect(wrapper.html()).toMatchSnapshot()
  })

  describe('while loading', () => {
    let wrapper: VueWrapper<typeof Button>
    beforeEach(() => {
      wrapper = shallowMount(Button, {
        props: { loading: true, loadingMsg: 'processing' },
      }) as any
    })
    it('show loading message', async () => {
      expect(wrapper.get('[aria-live="assertive"]').text()).toBe('processing')
    })

    it('should not be able to click', async () => {
      await wrapper.trigger('click')
      expect(wrapper.emitted()).not.toHaveProperty('click')
    })
  })

  it('emit click correctly', async () => {
    const wrapper = shallowMount(Button)
    await wrapper.trigger('click')
    expect(wrapper.emitted().click).toHaveLength(1)
  })

  it('should not be able to cick when aria-disabled is true', async () => {
    const wrapper = shallowMount(Button, {
      attrs: {
        'aria-disabled': true,
      },
    })
    await wrapper.trigger('click')
    expect(wrapper.emitted()).not.toHaveProperty('click')
  })
})
