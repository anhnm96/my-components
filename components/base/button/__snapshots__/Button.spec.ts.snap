// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button.vue > render correctly 1`] = `
"<button data-v-b4ce9510="" class="btn relative"><span data-v-b4ce9510="" class="inline-flex items-center"></span>
  <!--v-if-->
</button>"
`;
